packages/hyperion-autologging/src/ALSurfaceVisibilityPublisher.ts

```diff
@@ -119,29 +124,151 @@ export function publish(options: InitOptions): void {
+    setVirtualPropertyValue<ALSurfaceData | null>(
+      element,
+      SURFACE_DATA_PROP,
+      surfaceData,
+    )
+    observedRoots.set(element, surfaceData);
+    surfaceDataRoots.set(surfaceData, element);
+  }
+
+  function untrackElement(observer: IntersectionObserver, element: Element, surfaceData: ALSurfaceData): void {
+    observer.unobserve(element);
+    setVirtualPropertyValue<IntersectionObserver | null>(
+      element,
+      VISIBILITY_OBSERVER_PROP,
+      null,
+    )
+    setVirtualPropertyValue<ALSurfaceData | null>(
+      element,
+      SURFACE_DATA_PROP,
+      null,
+    )
+    observedRoots.delete(element, surfaceData);
+    surfaceDataRoots.delete(surfaceData, element);
+  }
+
+  let mutationObserver: MutationObserver | null = null;
+  if (enableDynamicChildTracking) {
+    mutationObserver = new MutationObserver((mutations) => {
+      for (const mutation of mutations) {
+        if (mutation.type === 'childList') {
+          const element = mutation.target as Element;
+          const observer = getVirtualPropertyValue<IntersectionObserver | null>(element, VISIBILITY_OBSERVER_PROP);
+          const surfaceData = getVirtualPropertyValue<ALSurfaceData | null>(element, SURFACE_DATA_PROP);
+          if (!observer || !surfaceData) {
+            return;
+          }
+          // Handle removed nodes
+          if (mutation.removedNodes.length > 0) {
+            handleRemovedNodes(observer, element, surfaceData, mutation.removedNodes);
+          }
+
+          // Handle added nodes
+          if (mutation.addedNodes.length > 0) {
+            // Re-evaluate the entire element with its new children
+            const surfaceDataList = observedRoots.get(element);
+            if (surfaceDataList && surfaceDataList.includes(surfaceData)) {
+              // Untrack the parent element first
+              untrackElement(observer, element, surfaceData);
+              // Try to find suitable children with the updated DOM
+              const roots = getNonSurfaceWrapperRoots(element);
+              if (roots.length > 0) {
+                // Found suitable children, observe them
+                for (let i = 0; i < roots.length; ++i) {
+                  trackElement(observer, roots[i], surfaceData);
+                  break; // Only need to observe one viable child
+                }
+              } else {
+                // Still no suitable children, track the parent again
+                // Which will check for observable roots again, and attach a mutation observer if needed
+                trackElement(observer, element, surfaceData);
+              }
+            }
+          }
+        }
+      }
+    });
+  }
+
+  // Handle an element by either observing its suitable children or setting up a mutation observer
+  function handleElementWithChildren(
+    observer: IntersectionObserver,
+    element: Element,
+    surfaceData: ALSurfaceData,
+    mutObserver: MutationObserver | null = mutationObserver,
```

Comment:
I only see two call sites to this function, and none of them pass any `mutObserver`. Did I miss something here?

Reply:
Yes I had started to apply your suggestion but noted that `break;` above.  Let me clean this up.